### PR TITLE
updateRule/the-best-dependency-injection-container

### PR DESCRIPTION
**Pull request by YazhiChen**
> 1. What triggered this change? (PBI link, Email Subject, conversation + reason, etc)

✏️
As per @adamcogan 's email with subject 'SugarLearning - Remove IoC' on Jan 16th, 2024, below PBIs were created. The IoC class and StructureMap have now been eliminated from the project, replaced by .NET's built-in Dependency Injection, which is the recommended DI solution over Autofac.

- [PBI 88033](https://dev.azure.com/ssw/SSW.Induction/_workitems/edit/88033): ♻️ Replace IoC/StructureMap with .NET 8 DI
- [PBI 89588](https://dev.azure.com/ssw/SSW.Induction/_workitems/edit/89588): 📝 Update rule/Do you know the best dependency injection container?

> 2. What was changed?

✏️
- Update the old image
- Suggest using .NET built-in DI over other options
- Provide examples of using .NET built-in DI for both web apps and console apps

> 3. Did you do pair or mob programming (list names)?

✏️I worked with @ozairashfaqueSSW

### CODE REVIEW

File: rules/the-best-dependency-injection-container/rule.md [hunk 1]
```diff
@@ -1,7 +1,7 @@
 ---
 type: rule
 archivedreason: 
-title: Do you know the best dependency injection container? (aka Do not waste days evaluating IOC containers)
+title: Do you know the best dependency injection container? 
 guid: d4fc76e9-0802-48bd-83d7-4e068a19d33b
 uri: the-best-dependency-injection-container
 created: 2013-02-01T16:43:44.0000000Z
```
Comment (danielmackay) [inline rules/the-best-dependency-injection-container/rule.md:7]: Add yourself to the list of authors

File: rules/the-best-dependency-injection-container/rule.md [hunk 2]
```diff
@@ -24,14 +24,12 @@ redirects:
 
 ---
 
-You can waste days evaluating IOC containers. The top ones are quite similar. There is not much in this, but the best ones are StructureMap and AutoFac. At SSW we use Autofac on most projects.
+IoC containers are powerful tools that implement the IoC principle and automate the process of dependency resolution and object instantiation. They serve as central repositories of services and manage the lifecycle of objects. Some popular IoC containers in the .NET ecosystem include StructureMap, Autofac, Castle Winsdor, Ninject, Unity. At SSW we recommend using [.NET built-in Dependency Injection](https://learn.microsoft.com/en-us/dotnet/core/extensions/dependency-injection) as default. Read more on [Dependency injection in ASP.NET Core](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/dependency-injection?view=aspnetcore-8.0).
```
Comment (danielmackay) [inline rules/the-best-dependency-injection-container/rule.md:27]: Can you move the options to a separate list and make it clear which one is recommended, which ones are neutral, and which ones are NOT recommended?

File: rules/the-best-dependency-injection-container/rule.md [hunk 3]
```diff
@@ -44,17 +42,69 @@ The top tools all contain comparable functionality. In practice which one you us
...
 ::: bad  
-![Figure: Bad Example - Ninject was a top container but is no longer developed as actively as Autofac and Structuremap. Both Autofac and Structuremap have active communities and contributors that ensure they stay up to date with the latest changes in .Net](dic-bad.png)  
+![Figure: Bad Example - Ninject and StructureMap were top containers but are no longer actively developed. Together with Autofac, they do not support the latest version of .NET](di-container-bad.png)  
+:::
+
+```csharp
+using SSW.SugarLearning.DependencyResolver;
+using StructureMap;
+
+namespace SSW.SugarLearning.WebJob.BadgeTask
```
Comment (danielmackay) [inline rules/the-best-dependency-injection-container/rule.md:52]: Make the namespace generic?  Or remove altogether?

File: rules/the-best-dependency-injection-container/rule.md [hunk 4]
```diff
@@ -44,17 +42,69 @@ The top tools all contain comparable functionality. In practice which one you us
...
+::: bad
+Bad example - Use the StructureMap IoC container but did not do the proper dependency injection
 :::
 
-::: good  
-![Figure: Good Example - Autofac has a great combination of performance and features and is actively developed](dic-good.png)  
+```csharp
+var builder = WebApplication.CreateBuilder(args);
+
+builder.Services.AddApplication();
```
Comment (danielmackay) [inline rules/the-best-dependency-injection-container/rule.md:75]: Can we have a simpler example here?  There's lots of code that's not relevant to IoC.

File: rules/the-best-dependency-injection-container/rule.md [hunk 5]
```diff
@@ -44,17 +42,69 @@ The top tools all contain comparable functionality. In practice which one you us
...
+::: good
+Good example - Use ASP.Net Core built-in Dependency Injection for web app
 :::
 
-**Note:** Autofac's support for child lifetime containers may be  significant for some: [An Autofac Lifetime Primer](http://nblumhardt.com/2011/01/an-autofac-lifetime-primer/).
+```csharp
+var builder = Host.CreateApplicationBuilder(args);
+IHostEnvironment env = builder.Environment;
+
+builder.Configuration
```
Comment (danielmackay) [inline rules/the-best-dependency-injection-container/rule.md:94]: Make this example simpler.

File: rules/the-best-dependency-injection-container/rule.md (hunk 2 above)
Comment (GordonBeeming) [inline rules/the-best-dependency-injection-container/rule.md:27]: We need to say what IoC is, before using it lots of times 🙂

Comment (danielmackay) [reply rules/the-best-dependency-injection-container/rule.md:27]: Similar to the options in: https://www.ssw.com.au/rules/use-mass-transit/

**Review by danielmackay — CHANGES_REQUESTED** [review]
Hi Yazhi.  A few comments to tidy up.  Look forward to seeing the changes.

File: rules/the-best-dependency-injection-container/rule.md (hunk 2 above)
Comment (YazhiChen) [reply rules/the-best-dependency-injection-container/rule.md]: Done.

File: rules/the-best-dependency-injection-container/rule.md (hunk 1 above)
Comment (YazhiChen) [reply rules/the-best-dependency-injection-container/rule.md:7]: Done.

File: rules/the-best-dependency-injection-container/rule.md (hunk 2 above)
Comment (YazhiChen) [reply rules/the-best-dependency-injection-container/rule.md]: Done.

File: rules/the-best-dependency-injection-container/rule.md (hunk 3 above)
Comment (YazhiChen) [reply rules/the-best-dependency-injection-container/rule.md]: Removed.

File: rules/the-best-dependency-injection-container/rule.md (hunk 4 above)
Comment (YazhiChen) [reply rules/the-best-dependency-injection-container/rule.md]: Removed irrelevant code

File: rules/the-best-dependency-injection-container/rule.md [hunk 6]
```diff
@@ -12,6 +12,8 @@ authors:
   url: https://ssw.com.au/people/adam-cogan
 - title: Brendan Richards
   url: https://ssw.com.au/people/brendan-richards
+  - title: Yazhi Chen
```
Comment (danielmackay) [inline rules/the-best-dependency-injection-container/rule.md:15]:
```suggestion
- title: Yazhi Chen
```

Comment (danielmackay) [reply rules/the-best-dependency-injection-container/rule.md:15]: Remove uneeded spaces

**Review by danielmackay — CHANGES_REQUESTED** [review]
One comment

**Review by danielmackay — APPROVED** [review]
LGTM :-)

**Pull request merged**